### PR TITLE
Kafka converter and row level policies

### PR DESCRIPTION
**Pull request by zliu41**
- Implemented AvroProjectionConverter for removing certain fields from an Avro schema or an Avro record.

- Added two row-level policies for Kafka avro records: one for timestamp and one for GUID.

### CODE REVIEW

File: gobblin-core/src/main/java/gobblin/converter/filter/AvroProjectionConverter.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+/**
+ * A {@link Converter} that removes certain fields from an Avro schema or an Avro record.
+ *
+ * @author ziliu
+ */
+public class AvroProjectionConverter extends AvroToAvroConverterBase {
+
+  public static final String REMOVE_FIELDS = ".remove.fields";
+
+  private AvroSchemaFieldRemover fieldRemover;
```
Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/converter/filter/AvroProjectionConverter.java]: You can use `Optional`.

File: gobblin-core/src/main/java/gobblin/converter/filter/AvroSchemaFieldRemover.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    this.addChildren(fieldNames);
+  }
+
+  private AvroSchemaFieldRemover() {
+    this.children = Maps.newHashMap();
+    this.schemaMap = Maps.newHashMap();
+  }
+
+  private void addChildren(String fieldNames) {
+    for (String fieldName : Splitter.on(',').trimResults().omitEmptyStrings().splitToList(fieldNames)) {
```
Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/converter/filter/AvroSchemaFieldRemover.java]: You can make the `Splitter`s here as constants.

File: gobblin-core/src/main/java/gobblin/converter/filter/AvroSchemaFieldRemover.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+      List<String> fieldNameComponents = Splitter.on('.').trimResults().omitEmptyStrings().splitToList(fieldName);
+      if (!fieldNameComponents.isEmpty()) {
+        this.addChildren(fieldNameComponents, 0);
+      }
+    }
+  }
+
+  private void addChildren(List<String> fieldNameComponents, int level) {
+    if (!this.children.containsKey(fieldNameComponents.get(level))) {
```
Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/converter/filter/AvroSchemaFieldRemover.java]: `containsKey` may throw an NPE if the key is null, which may be the case if `fieldNameComponents` does not contain `level`.

File: gobblin-core/src/main/java/gobblin/converter/filter/AvroSchemaFieldRemover.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        return this.removeFieldsFromMap(schema);
+      default:
+        return schema;
+    }
+  }
+
+  private Schema removeFieldsFromRecords(Schema schema) {
+    List<Field> newFields = Lists.newArrayList();
+    for (Field field : schema.getFields()) {
+      if (this.shouldRemove(field) == false) {
```
Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/converter/filter/AvroSchemaFieldRemover.java]: `if (!this.shouldRemove(field))`.

File: gobblin-core/src/main/java/gobblin/converter/filter/AvroSchemaFieldRemover.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    Schema newRecord = Schema.createRecord(schema.getName(), schema.getDoc(), schema.getNamespace(), schema.isError());
+    newRecord.setFields(newFields);
+    this.schemaMap.put(schema.getFullName(), newRecord);
+    return newRecord;
+  }
+
+  private boolean shouldRemove(Field field) {
+    return this.children.containsKey(field.name()) && this.children.get(field.name()).children.isEmpty();
```
Comment (liyinan926) [inline gobblin-core/src/main/java/gobblin/converter/filter/AvroSchemaFieldRemover.java]: Can you add comment on the logic here?

File: gobblin-utility/src/main/java/gobblin/util/AvroUtils.java [hunk 6]
```diff
@@ -108,4 +121,29 @@
...
+   * Change the schema of an Avro record.
+   * @param record The Avro record whose schema is to be changed.
+   * @param newSchema The target schema. It must be compatible as reader schema with record.getSchema() as writer schema.
+   * @return a new Avro record with the new schema.
+   * @throws IOException if conversion failed.
+   */
+  public static GenericRecord convertRecordSchema(GenericRecord record, Schema newSchema) throws IOException {
+    Preconditions.checkArgument(checkReaderWriterCompatibility(newSchema, record.getSchema()).getType() == COMPATIBLE);
+
+    ByteArrayOutputStream out = new ByteArrayOutputStream();
```
Comment (liyinan926) [inline gobblin-utility/src/main/java/gobblin/util/AvroUtils.java]: Please use a `Closer` to close the output stream and other `Closeable`s.

**Comment by zliu41** [comment]
Updated.

**Comment by liyinan926** [comment]
LGTM.

**Pull request merged**